223 ongoing events

pferdefleisch opened the pull request:
Added ongoing events. Take a look, especially at the Translation commit. I will add tests after we OK this. Cheers :kiss:

klappradla commented on app/controllers/ngos/ongoing_events_controller.rb:
```diff
@@ -0,0 +1,74 @@
+      current_ngo: current_ngo,
+      ongoing_event: event_params(params)
+    ).model
+    if @event.valid?
```
Won't it already crash within `process` for invalid events? I don't know if you'd even reach this statement...

pferdefleisch replied:
No, I'm using `.create` which just gives you an invalid object back.

klappradla replied:
A sure, it does the rollback itself and gives you the invalid object - just tried it out myself. Totally forgot 😉

klappradla commented on app/controllers/ngos/ongoing_events_controller.rb:
```diff
@@ -0,0 +1,74 @@
+        event_id: params[:id],
+        ongoing_event: event_params(params)
+      ).model
+    if @event.valid?
```
Same as for `create`: do you even reach this statement when passing invalid `event_params`?

pferdefleisch replied:
same for `.update_attributes` (see above)

groe commented:
The "Events" item in the layouts navigation bar is now inactive if you navigate between Einsätze and Dauerhafte Einsätze.

Maybe `GET /` could redirect to `GET /events` (instead of rendering the list) so we only have 1 route for `events#index`?

klappradla commented on config/routes.rb:
```diff
@@ -55,7 +59,12 @@
-  resources :events, only: [:index, :show]
+  resources :events,         only: [:index, :show]
+  resources :ongoing_events, only: [:index, :show] do
+    post   :opt_in,  on: :member
```
Would this be a candidate for a concern?
```ruby
concern :participation do
  post :opt_in
  delete :opt_out
end

....

resources :ongoing_events, only: [:index, :show], concerns: :participation
resources :shifts, only: [:show], concerns: :participation
```

Just came to my mind, rare occasion where this could be useful.

pferdefleisch replied:
If we have 3.

klappradla commented on app/models/ongoing_event.rb:
```diff
@@ -0,0 +1,43 @@
+  scope :newest_first, -> { order(:created_at) }
+  scope :published,    -> { where.not(published_at: nil) }
+
+  def pending?
```
If the whole `state` handling is the same for `event` and `ongoing_event`, it would eventually make sense to extract it into a concern and only test it once or with a shared example. Might go out of hand pretty quick otherwise.

klappradla replied:
Besides that, if it makes sense to implement it in another way, totally go for it. I just put it like this, cause it was behaviorwise the closest I could get from refactoring some models using state machines a day before the release.

pferdefleisch replied:
I'd rather wait. If we start adding things and it gets ugly, or if there is a third thing, ok. I rarely do this kind of abstraction on only two things.

groe commented:
There are currently no english translations.

klappradla commented:
@pferdefleisch I'm don't really have time now to test the code locally, so I just left added the review. Looks fine all in all 🚀 

Slight doubts about raising exceptions for invalid events though - I just don't understand how it should work. But if it does, 💪 

Translations look fine as well.

pferdefleisch commented:
@groe This is ready to go. You can deploy it after you merge it. I will have the data migration ready tomorrow.

groe commented:
The personal schedule at "Deine Einsätze / Bevorstehende" seems to list *all* the events, not only the ones I have signed up for?

groe commented:
@pferdefleisch I've updated my comment above

pferdefleisch commented:
@groe For your three points:

1. Either I don't understand your problem or I am not seeing the same behavior. I think you mean - if i am signed in as a volunteer and I go to the `Einsätze` link -> then I click on `Dauerhafte` -> you are expecting to see only the ongoing events you are signed up for, but instead you are seeing all of the ongoing events. I am seeing only the ones I am signed up for.

2. I have added the counts to the ngo/ongoing events/show page. It is there just to remind the ngo of how many volunteers they thought they needed. Spoke with Michael about this, but totally forgot to add it to the show page.

3. Volunteers are now visible with an email link even ( :open_mouth: )

Just add more things if you need them and please clarify 1.

groe commented:
Thanks @pferdefleisch!

Re 1: Yes the `Dauerhafte` link works correctly but `Bevorstehende` does not (not sure if this is related to this PR though). I had my DB cleared and re-populated just before I tested this but there were dozens of events in `Bevorstehende` already when I logged in for the first time, and when I clicked on them the event page said I was not in fact not signed up. But then it shouldn't even appear in the list, or?

pferdefleisch commented:
@groe This was a bug I introduced in another pull request. Fixed in https://github.com/where2help/where2help/pull/244/commits/6789b0fea3447b4699f634858be40b347280f3c4
Are we ready to roll?

groe commented:
Perfect, thank you @pferdefleisch
